src/Lean/Server/Rpc/Basic.lean

```diff
@@ -33,36 +40,98 @@ end Lean.Lsp
 
 namespace Lean.Server
 
-structure RpcObjectStore : Type where
-  /-- Objects that are being kept alive for the RPC client, together with their type names,
-  mapped to by their RPC reference.
+/--
+Marks values to be encoded as opaque references in RPC packets.
+Two `WithRpcRef`s with the same `id` will yield the same `RpcRef`.
+
+See also the docstring for `RpcEncodable`.
+-/
+structure WithRpcRef (α : Type u) where
+  private mk' ::
+    val : α
+    id  : USize
+  deriving Inhabited
 
-  Note that we may currently have multiple references to the same object. It is only disposed
-  of once all of those are gone. This simplifies the client a bit as it can drop every reference
-  received separately. -/
-  aliveRefs : PersistentHashMap Lsp.RpcRef Dynamic := {}
-  /-- Value to use for the next `RpcRef`. It is monotonically increasing to avoid any possible
-  bugs resulting from its reuse. -/
-  nextRef   : USize := 0
+builtin_initialize freshWithRpcRefId : IO.Ref USize ← IO.mkRef 0
 
-def rpcStoreRef (any : Dynamic) : StateM RpcObjectStore Lsp.RpcRef := do
-  let st ← get
-  set { st with
-    aliveRefs := st.aliveRefs.insert ⟨st.nextRef⟩ any
-    nextRef := st.nextRef + 1
-  }
-  return ⟨st.nextRef⟩
+/--
+Creates an `WithRpcRef` instance with a unique `id`.
+Two `WithRpcRef`s with the same `id` will yield the same `RpcRef`.
+Hence, storing a `WithRpcRef` produced by `WithRpcRef.mk` and serving it to the client twice will
+also yield the same `RpcRef` twice, allowing clients to reuse their associated UI state across
+RPC requests.
```

Comment:
```suggestion
As long as the client holds at least one reference to this `WithRpcRef`,
serving it again will yield the same client-side reference.
Thus, when used as React deps,
client-side references can help preserve UI state across RPC requests.
```
I tried to clarify that at least RC=1 on the server is necessary to actually serve the same `RpcRef`, and suggested using as React dep explicitly.